src/core/algorithms/nd/nd_verifier/nd_verifier.cpp

```diff
@@ -0,0 +1,228 @@
+                                   return std::move(a) + ", " + std::to_string(b);
+                               });
+    };
+
+    LOG(INFO) << "Parameters of NDVerifier:";
+    LOG(INFO) << "\tInput table: " << input_table_->GetRelationName();
+    LOG(INFO) << "\tNull equals null: " << is_null_equal_null_;
+    LOG(INFO) << "\tLhs indices: " << indices_to_str(lhs_indices_);
+    LOG(INFO) << "\tRhs indices: " << indices_to_str(rhs_indices_);
+    LOG(INFO) << "\tWeight: " << weight_;
+
+    VerifyND();
+
+    auto elapsed_milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now() - start_time);
+
+    LOG(WARNING) << "NDVerifier::ExecuteInternal finished in "
+                 << std::to_string(elapsed_milliseconds.count())
+                 << "ms";  // We use std::to_string, because compiler on github doesn`t
+                           // have implementation for stringstream::operator<<(unsigned)
+
+    CalculateStats();
+
+    elapsed_milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::system_clock::now() - start_time);
+
+    return elapsed_milliseconds.count();
+}
+
+bool NDVerifier::NDHolds() const {
+    return stats_calculator_.GetRealWeight() <= weight_;
+}
+
+void NDVerifier::ResetState() {
+    stats_calculator_ = util::StatsCalculator{};
+}
+
+std::pair<std::shared_ptr<std::vector<std::string>>, std::shared_ptr<std::vector<size_t>>>
+NDVerifier::EncodeMultipleValues(config::IndicesType const& col_idxs) const {
+    auto values = std::make_shared<std::vector<std::string>>();
+    auto row = std::make_shared<std::vector<size_t>>();
+
+    for (size_t row_idx{0}; row_idx < typed_relation_->GetNumRows(); ++row_idx) {
+        std::stringstream ss;
+        ss << '(';
+        bool is_null = false;
+        for (auto col_idx_pt{col_idxs.begin()}; col_idx_pt != col_idxs.end(); ++col_idx_pt) {
+            auto const& col_data = typed_relation_->GetColumnData(*col_idx_pt);
+            auto const& byte_data = col_data.GetData();
+            auto const& type = col_data.GetType();
+
+            if (col_idx_pt != col_idxs.begin()) {
+                ss << ", ";
+            }
+
+            auto const* bytes_ptr = byte_data[row_idx];
+            if (bytes_ptr == nullptr) {
+                LOG(INFO) << "WARNING: Cell (" << *col_idx_pt << ", " << row_idx << ") is empty";
+                is_null = true;
+            } else {
+                ss << type.ValueToString(bytes_ptr);
+            }
+        }
+        ss << ')';
+
+        auto string_data = ss.str();
+        size_t index{values->size()};
+        if (is_null && !is_null_equal_null_) {
+            // If not null_eq_null, every value with null will be unique
+            values->push_back(string_data);
+        } else {
+            for (size_t i{0}; i < values->size(); ++i) {
+                if (values->at(i) == string_data) {
```

Comment:
- Method .at() provides bound check, but this can be useful during the implementation of the algorithm to catch errors during debugging. In other situations we don't want to have overhead for these checks, so it is preferable to use operator[]. In theory, method .at() can be used when the index comes from the user-side to avoid undefined behavior in case of an invalid index. However, in this case it is preferable to use special checks and throw special exceptions.
- The previous arguments are also true for smart pointers. In this case, you should write (*ptr)[index] instead of ptr->operator[](index). If the expression is very long, you can define additional variable (e.g. reference to vector in shared_ptr) and use it without pointer dereference.